pkg/wrangler/capi.go

```diff
@@ -2,44 +2,65 @@ package wrangler
 
 import (
 	"context"
+	"fmt"
 	"sync"
-	"time"
+	"sync/atomic"
 
 	capi "github.com/rancher/rancher/pkg/generated/controllers/cluster.x-k8s.io"
+	wapiextv1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/apiextensions.k8s.io/v1"
 	"github.com/rancher/wrangler/v3/pkg/generic"
 	"github.com/sirupsen/logrus"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// ManageDeferredCAPIContext polls for the availability of CAPI CRDs and registers deferred controllers
+// manageDeferredCAPIContext polls for the availability of CAPI CRDs and registers deferred controllers
 // and executes deferred functions once they are available. Once CAPI CRDs are found, this function will
 // not continue polling. Individual registration calls can be made once polling is complete by directly using
 // the DeferredCAPIRegistration struct.
-func (w *Context) ManageDeferredCAPIContext(ctx context.Context) {
-	ticker := time.NewTicker(5 * time.Second)
-	defer ticker.Stop()
+func (w *Context) manageDeferredCAPIContext(ctx context.Context) {
+	logrus.Info("[deferred-capi - manageDeferredCAPIContext] Starting to monitor CAPI CRD availability")
+	var done atomic.Bool
+	w.CRD.CustomResourceDefinition().OnChange(ctx, "capi-deferred-registration", func(key string, crd *apiextv1.CustomResourceDefinition) (*apiextv1.CustomResourceDefinition, error) {
+		if done.Load() {
+			return crd, nil
+		}
 
-	logrus.Debug("[deferred-capi] Starting to monitor CAPI CRD availability")
+		if !capiCRDsReady(w.CRD.CustomResourceDefinition().Cache()) {
+			return crd, nil
+		}
 
-	for {
-		allCRDsReady := w.checkCAPICRDs()
-		if allCRDsReady {
-			logrus.Debug("[deferred-capi] All CAPI CRDs are now available and established.")
-			w.createCAPIFactoryAndStart(ctx)
-			return
+		if !done.CompareAndSwap(false, true) {
+			return crd, nil
 		}
 
-		select {
-		case <-ctx.Done():
-			logrus.Error("[deferred-capi] Context cancelled while waiting for CAPI CRDs")
-			return
-		case <-ticker.C:
+		logrus.Info("[deferred-capi - manageDeferredCAPIContext] initializing CAPI factory")
+		opts := &generic.FactoryOptions{
+			SharedControllerFactory: w.ControllerFactory,
 		}
-	}
+
+		capi, err := capi.NewFactoryFromConfigWithOptions(w.RESTConfig, opts)
+		if err != nil {
+			logrus.Fatalf("Encountered unexpected error while creating capi factory: %v", err)
+		}
+		w.DeferredCAPIRegistration.mutex.Lock()
+		defer w.DeferredCAPIRegistration.mutex.Unlock()
```

Comment:
Q: If we need a lock that guards `w.capi` and `w.CAPI` shouldn't we have the mutex on `w` rather than `w.DeferredCAPIRegistration`?

Reply:
If w.capi has to be guarded it will be another lock, not a move of the `w.DeferredCAPIRegistration` lock into `w`. That lock looks to be needed too.